### PR TITLE
Added support for templates, with each template for now only containing user definitions

### PR DESCRIPTION
**Pull request by techsmyth**
new entity: Template
new query: templates
new mutation: createTemplate
new sample file for creating a new template

The population of a template to be another issue.

### CODE REVIEW

File: src/domain/ecoverse/ecoverse.service.ts [hunk 1]
```diff
@@ -159,6 +163,17 @@ export class EcoverseService {
     }
   }
 
+  async getTemplates(): Promise<ITemplate[]> {
+    try {
+      const ecoverseId = await this.getEcoverseId();
+      const templates = await this.templateService.getTemplates(ecoverseId);
+      return templates;
+    } catch (e) {
+      // this.eventDispatcher.dispatch(events.logger.error, { message: 'Something went wrong in getMembers()!!!', exception: e });
```
Comment (valentinyanakiev) [inline src/domain/ecoverse/ecoverse.service.ts]: Delete this line.

File: src/domain/ecoverse/ecoverse.service.ts [hunk 2]
```diff
@@ -159,6 +163,17 @@ export class EcoverseService {
...
+      throw e;
```
Comment (valentinyanakiev) [inline src/domain/ecoverse/ecoverse.service.ts]: No point to rethrow, with the entry above commented / removed and no context added to the Error object, this is pointless. Unhandled exceptions are handled by Nest anyway.

File: src/domain/template/template.entity.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  @Field(() => [User], {
+    nullable: true,
+    description:
+      'The set of user types that are available within this template',
+  })
+  @ManyToMany(
+    () => User,
+    user => user.templates,
+    { eager: true, cascade: true }
```
Comment (valentinyanakiev) [inline src/domain/template/template.entity.ts:38]: Many-to-many and eager loading will be a performance disaster most likely.

File: src/domain/template/template.entity.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  )
+  @JoinTable({ name: 'user_group_members' })
```
Comment (valentinyanakiev) [inline src/domain/template/template.entity.ts]: You can't use the same join table as the one defining the user-group relationship, this needs to be a separate join table for the many-to-many.

**Review by valentinyanakiev — CHANGES_REQUESTED** [review]
Several changes required. Also, have a look at the build https://travis-ci.com/github/cherrytwist/Server/builds/197020078. The changes to the model require changes to the tests, otherwise, the whole pipeline will start failing.

File: src/domain/template/template.entity.ts (hunk 3 above)
Comment (techsmyth) [reply src/domain/template/template.entity.ts:38]: Only going to be one user near term, so will not have an impact imho...

File: src/domain/template/template.entity.ts (hunk 4 above)
Comment (techsmyth) [reply src/domain/template/template.entity.ts]: good catch - missed this one. Will fix...

File: src/domain/template/template.entity.ts (hunk 3 above)
Comment (techsmyth) [reply src/domain/template/template.entity.ts:38]: First lets get it working, will be very little in the template for now

File: src/domain/ecoverse/ecoverse.service.ts (hunk 2 above)
Comment (techsmyth) [reply src/domain/ecoverse/ecoverse.service.ts]: Agreed, removed all the catching / rethrowing which was from a while back

**Comment by techsmyth** [comment]
Think this is good to go - I cannot reproduce the test results locally, will see what happens with the new commit...

**Review by valentinyanakiev — APPROVED** [review]

**Pull request merged**